src/main/scala/nl/knaw/dans/api/sword2/DepositHandler.scala

```diff
@@ -152,54 +166,157 @@ object DepositHandler {
+          listIterator().asScala.
+          filterNot(msg => msg.getCode == CompleteVerifier.CODE_PAYLOAD_MANIFEST_CONTAINS_MISSING_FILE)
+        if (otherThanMissingPayloadFilesMessages.nonEmpty)
+          Failure(InvalidDepositException(id, otherThanMissingPayloadFilesMessages.mkString))
+        else {
+          val missingPayloadFiles = validationResult.getSimpleMessages.
+            listIterator().asScala.toList.
+            flatMap(_.getObjects.asScala.toList)
+          val fetchItemFilesBagStore = fetchItemsBagStore.map(_.getFilename)
+          val missingFilesNotInFetchText = missingPayloadFiles.filterNot(file => fetchItemFilesBagStore.contains(file))
+          if (missingFilesNotInFetchText.nonEmpty)
+            Failure(InvalidDepositException(id, s"Missing payload files not in the fetch.txt: ${missingFilesNotInFetchText.mkString}."))
+          else {
+            if (noFetchItemsAlreadyInBag(bagitDir, fetchItemsBagStore).isFailure)
+              Failure(InvalidDepositException(id, s"One or more fetch.txt files are already present in the bag."))
+            else
+              validateChecksumsFetchItems(bag, fetchItemsBagStore)
+          }
+        }
+      }
+    }
+  }
+
+  private def resolveFetchItems(bagitDir: File, fetchItems: Seq[FetchTxt.FilenameSizeUrl])(implicit id: String): Unit = {
+    if (fetchItems.nonEmpty) {
+      log.debug(s"[$id] Resolving files in fetch.txt, those referring outside the bag store.")
+      fetchItems.foreach(item =>
+        Using.urlInputStream(new URL(item.getUrl)).foreach(src => {
+          val file = new File(bagitDir.getAbsoluteFile, item.getFilename)
+          if (file.exists()) {
+            Failure(InvalidDepositException(id, s"File ${item.getFilename} in the fetch.txt is already present in the bag."))
+          } else {
+            file.getParentFile.mkdirs()
+            Files.copy(src, file.toPath)
+          }
+        }))
+    }
+  }
+
+  private def noFetchItemsAlreadyInBag(bagitDir: File, fetchItems: Seq[FetchTxt.FilenameSizeUrl])(implicit id: String): Try[Unit] = {
+    log.debug(s"[$id] Checking that the files in fetch.txt are absent in the bag.")
+
+    val presentFiles = fetchItems.filter(item => new File(bagitDir.getAbsoluteFile, item.getFilename).exists())
+    if (presentFiles.nonEmpty)
+      Failure(InvalidDepositException(id, s"Fetch.txt file ${presentFiles.head.getFilename} is already present in the bag."))
+    else
+      Success(Unit)
+  }
+
+  private def validateChecksumsFetchItems(bag: Bag, fetchItems: Seq[FetchTxt.FilenameSizeUrl]) (implicit id: String, baseDir: File, baseUrl: URI): Try[Unit] = {
+    log.debug(s"[$id] Validating checksums of those files in fetch.txt, that refer to the bag store.")
+
+    val fetchItemFiles = fetchItems.map(_.getFilename)
+    val checksums = bag.getPayloadManifests.asScala.toList.
+                      flatMap(_.asScala).
+                      filter({case (file, checksum) =>
```

Comment:
As you said, this already looks so much better indeed. Two pieces of follow-up:

1. You don't need the `()` here, now that you use `{ case ... }` (saw a similar thing in other places, btw)
2. You don't use `checksum`, so you can write the pattern match as `case (file, _)`. That way it is clear to everybody that you only use the first part of the tuple and are not interested in the second part for now...